docs/source/developers/websocket-protocols.rst

```diff
@@ -1,12 +1,108 @@
+appropriate logic around kernel shutdown, like asking the kernel to shut down
+first through a kernel message, then forcefully shutting down the kernel if
+there is no response.
+
+Kernel messages
+---------------
+
+A websocket client connecting to a kernel through the Jupyter server websocket
+bridge will use messages according to the Jupyter kernel message spec, with the
+modifications noted below.
+
+Kernel Message Specification Changes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+**Message Channels**
+
+The notebook server multiplexes all kernel message channels into a single
+websocket channel and encodes the channel name in the websocket message wire
+format (see below).
+
+**Kernel Status**
+
+The Jupyter server sends several additional `kernel status <https://jupyter-client.readthedocs.io/en/stable/messaging.html#kernel-status>`__ messages in addition
+to the kernel status messages that are sent by the kernel itself:
+
+1. When a kernel is restarted, an ``execution_state: 'restarting'`` kernel status message is sent.
+2. When a kernel dies, an ``execution_state: 'dead'`` kernel status message is sent.
+
+These status messages will have a different message header ``session`` value
+than the message header session values in messages from the kernel.
+
+.. note::
+
+    As an implementation detail, the message header session value of these
+    messages matches the `session_id` in the kernel connection URL.
+
+.. note::
+
+    In the classic notebook and JupyterLab client code, the websocket connection
+    is closed when explicitly requesting a restart or shutdown, so the
+    restarting and dead messages aren't received if it was requested by the
+    user. In those cases, receiving a ``restarting`` or ``dead`` message from
+    the notebook server means that the kernel had something happen to it, and
+    the user should be explicitly notified.
+
+**IOpub Message Rate Limits**
+
+The notebook server inspects the messages coming from the kernel to the client
+to rate-limit iopub messages. These rate limits can be raised.
+
+
+Buffering
+~~~~~~~~~
+
+If all websocket clients have disconnected from a kernel, the notebook server
+will temporarily buffer messages from the kernel to be delivered to the first
+websocket client that connects to the kernel.
+
+
+
+.. note::
+
+    In the classic notebook client and JupyterLab, requesting a kernel restart
+    immediately closes all websocket connections to the kernel, so kernel
+    buffering starts. When a new websocket connection is created connecting to
+    the kernel, the notebook server transmits all of the messages buffered from
+    the kernel. For the IPython kernel, this means the new websocket connection
+    will start with receiving status busy, shutdown_reply, and status idle
+    messages on the iopub channel from before the restart.
+
+.. note::
+
+    TODO
+
+    Document the session URL parameter used in kernel connections. Is that
```

Comment:
From @minrk at https://github.com/jupyter/notebook/pull/3765#discussion_r203521190

> 

Yes, we should document it. sessions_id is associated with each specific client. It's entirely unrelated to the session rest API. The session_id here is the message protocol header.session field, which is used on replies and iopub to identify the original sender of the message. The main important use of this session_id is setting zmq.IDENTITY on reconnects to ensure that replies intended for this client still arrive there after reconnects. It's also used to build messages created in the notebook server for this client, but I'm not sure we should commit to or rely on that yet.